flink-connectors/flink-hbase/src/main/java/org/apache/flink/addons/hbase/HBaseTableSourceInputFormat.java

```diff
@@ -0,0 +1,322 @@
+
+		resultScanner = table.getScanner(scan);
+		endReached = false;
+		scannedRows = 0;
+	}
+
+	private void logSplitInfo(String action, TableInputSplit split) {
+		int splitId = split.getSplitNumber();
+		String splitStart = Bytes.toString(split.getStartRow());
+		String splitEnd = Bytes.toString(split.getEndRow());
+		String splitStartKey = splitStart.isEmpty() ? "-" : splitStart;
+		String splitStopKey = splitEnd.isEmpty() ? "-" : splitEnd;
+		String[] hostnames = split.getHostnames();
+		LOG.info("{} split (this={})[{}|{}|{}|{}]", action, this, splitId, hostnames, splitStartKey, splitStopKey);
+	}
+
+	@Override
+	public boolean reachedEnd() throws IOException {
+		return endReached;
+	}
+
+	@Override
+	public Row nextRecord(Row reuse) throws IOException {
+		if (resultScanner == null) {
+			throw new IOException("No table result scanner provided!");
+		}
+		try {
+			Result res = resultScanner.next();
+			if (res != null) {
+				scannedRows++;
+				lastRow = res.getRow();
+				return mapResultToRow(res);
+			}
+		} catch (Exception e) {
+			resultScanner.close();
+			//workaround for timeout on scan
+			LOG.warn("Error after scan of " + scannedRows + " rows. Retry with a new scanner...", e);
+			scan.setStartRow(lastRow);
+			resultScanner = table.getScanner(scan);
+			Result res = resultScanner.next();
+			if (res != null) {
+				scannedRows++;
+				lastRow = res.getRow();
+				return mapResultToRow(res);
+			}
+		}
+		endReached = true;
+		return null;
+	}
+
+	private Row mapResultToRow(Result res) {
+		Object[] values = new Object[fieldNames.length];
+		int i = 0;
+		for(String field : fieldNames) {
+			String[] famCol = field.split(COLON);
+			byte[] value = res.getValue(Bytes.toBytes(famCol[0]), Bytes.toBytes(famCol[1]));
+			TypeInformation typeInfo = fieldTypeInfos[i];
+			if(typeInfo.isBasicType()) {
+				if(typeInfo.getTypeClass() ==  Integer.class) {
+					values[i] = Bytes.toInt(value);
+				} else if(typeInfo.getTypeClass() == Short.class) {
+					values[i] = Bytes.toShort(value);
+				} else if(typeInfo.getTypeClass() == Float.class) {
+					values[i] = Bytes.toFloat(value);
+				} else if(typeInfo.getTypeClass() == Long.class) {
+					values[i] = Bytes.toLong(value);
+				} else if(typeInfo.getTypeClass() == String.class) {
+					values[i] = Bytes.toString(value);
+				} else if(typeInfo.getTypeClass() == Byte.class) {
+					values[i] = value[0];
+				} else if(typeInfo.getTypeClass() == Boolean.class) {
+					values[i] = Bytes.toBoolean(value);
+				} else if(typeInfo.getTypeClass() == Double.class) {
+					values[i] = Bytes.toDouble(value);
+				} else if(typeInfo.getTypeClass() == BigInteger.class) {
+					values[i] = new BigInteger(value);
+				} else if(typeInfo.getTypeClass() == BigDecimal.class) {
+					values[i] = Bytes.toBigDecimal(value);
+				} else if(typeInfo.getTypeClass() == Date.class) {
+					values[i] = new Date(Bytes.toLong(value));
+				}
+			} else {
+				// TODO for other types??
```

Comment:
I can see that in the TableSourceITCase that uses CsvInputFormat - the metadata is also passed as the input and the meta data is registered with the table. Now in cases like hbase - how should that be done?